ARCs/arc-0037.md

```diff
@@ -0,0 +1,77 @@
+---
+arc: 37
+title: ABI Versioning
+description: Add a versioning scheme to the ABI
+author: Joe Polny (@joe-p)
+discussions-to: https://github.com/algorandfoundation/ARCs/issues/210
+status: Draft
+type: Standards Track
+category: ARC
+created: 2023-07-15
+requires: 4
+---
+
+## Abstract
+The Algorand ABI is defined in [ARC-0004](./arc-0004.md). As the protocol and desires of the ABI changes, there will be changes to the ABI. As such, there should be a versioning number to make it clear what versions of the ABI smart contracts adhere to and what versions of the ABI off-chain code supports.
+
+## Motivation
+Without any versioning scheme, it will be very hard to follow the evolution of the Algorand ABI.
+
+## Specification
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+Every ABI change is versioned `MAJOR.MINOR`. `MAJOR` is incremeneted when there are breaking changes and `MINOR` is incremented for the addition of backwards-compatible features.
+
+Any ABI Interface or Contract JSON description that adheres to ABI standards beyond [ARC-0004](./arc-0004.md) **MUST** inlcude a `abi-version` key in the top-level of the interface. Interface or Contract JSON descriptions that only adhere to [ARC-0004](./arc-0004.md) **SHOULD** have `"abi-version": "1.0"` at the top level, but descriptions without a `abi-version` key **MUST** be interepreted to implcitly be version 1.0.
```

Comment:
I wonder if a new key in the interface or contract description is enough. Right now if you want to call another contract's method in PyTeal, you only supply the app ID and the method signature (see [InnerTxnBuilder.MethodCall](https://pyteal.readthedocs.io/en/v0.25.0/api.html#pyteal.InnerTxnBuilder.MethodCall)). The SDKs and goal similarly only rely on method descriptions/signatures and app IDs.

Perhaps for major versions, it makes sense to somehow include this information in the method signature, e.g. `v2:foo()void`? This way, if an app I call decides to update its ABI version, my calls would start failing due to an unsupported method rather than one of us incorrectly decoding values. And if necessary, an app _could_ support the same method across multiple versions.

Reply:
One potential problem is that this realistically would have to be managed by the developer per method when working with high-level languages.

For example, let's say a developer writes a contract with the method `foo(uint64)void` in accordance with ARC-0004. Then an ARC like ARC-0043 (#214) is finalized inducing a major version bump and affects other methods in the contract and the developer wants to use the new ABI for their methods. This means the high-level language would use ABI 2.0, but does that mean it should break all existing clients that call `foo(uint64)void` even though the method is completely unaffected by the version change?

I suppose the high-level language could have logic to determine if the change is breaking or not, but seems a bit complex.

Reply:
Since you mentioned it (but part of other ARCS), I think the method for resolution need to be decorated same as in any HLL at build time for RT disambiguation.
To start, methods should probably be namespaced by ARC: otherwise we can't have contracts that implement multiple ARCs.  Based on method names being used in some example ARCs, chance of collision is *extremely* high.
They should also be namespaced by ABI version - otherwise calling convention changes will obv. break things.
so isOwner()void really becomes ARCXXXX/ABIX.X/isOwner()void pre-hash [for example purposes only]

Reply:
What if instead of versioning the full signature only affected types are versioned, with no version implicitly meaning 1.0. 

For example: `foo(string, uint64)void` -> `foo(v2:string,uint64)void`

Reply:
@joe-p interesting! That idea implies that there are no longer any backwards incompatible changes, since existing types are never changed, only antiquated by newer types.

Need to think through the implications of that more, but it seems promising.

Reply:
One other benefit to this is that you can have a method that takes in both v1 and v2 arguments. Perhaps you want to use v2 arguments internally in your methods logic but you need to pass some argument to an inner txn app call that only supports v1. So you could have a method like `foo(string, v2:string)void`. For something like ARC-0037 it's probably not needed since it's a matter of a few opcodes to convert between the two, but for future version that could have more complex changes this would be really useful for composability.

Reply:
I'm confused by the foo(string, v2:string) example.  Why would you have the same method use different encodings for two different arguments of the same type?
I would think having both:
foo(string, string)void
foo(v2:string, v2:string)void 
is what you'd see.

Reply:
> Perhaps you want to use v2 arguments internally in your methods logic but you need to pass some argument to an inner txn app call that only supports v1.

If you can't have both in one method, then you are forced to use whatever the lowest version supported by a potential inner txn app call is.

Reply:
You can accept arguments in any encoding you'd like, and then encode them differently for an inner call.

Reply:
Sure, but that could be costly for the opcode budget.